data/src/test/java/org/akvo/flow/data/repository/FormDataRepositoryTest.java

```diff
@@ -0,0 +1,245 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Akvo Flow is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.akvo.flow.data.repository;
+
+import android.database.Cursor;
+
+import org.akvo.flow.data.datasource.DataSourceFactory;
+import org.akvo.flow.data.datasource.DatabaseDataSource;
+import org.akvo.flow.data.datasource.files.FileDataSource;
+import org.akvo.flow.data.entity.ApiFormHeader;
+import org.akvo.flow.data.entity.FormHeaderParser;
+import org.akvo.flow.data.entity.FormIdMapper;
+import org.akvo.flow.data.entity.XmlParser;
+import org.akvo.flow.data.net.RestApi;
+import org.akvo.flow.data.net.s3.AmazonAuthHelper;
+import org.akvo.flow.data.util.ApiUrls;
+import org.akvo.flow.domain.util.DeviceHelper;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.DateFormat;
+import java.text.FieldPosition;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import io.reactivex.Observable;
+import io.reactivex.observers.TestObserver;
+import okhttp3.ResponseBody;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class FormDataRepositoryTest {
+
+    @Mock
+    FormHeaderParser mockFormHeaderParser;
+
+    @Mock
+    XmlParser mockXmlParser;
+
+    @Mock
+    DatabaseDataSource mockDatabaseDataSource;
+
+    @Mock
+    FileDataSource mockFileDataSource;
+
+    @Mock
+    FormIdMapper mockFormIdMapper;
+
+    @Mock
+    ApiFormHeader mockApiFormHeader;
+
+    @Mock
+    DeviceHelper mockDeviceHelper;
+
+    @Mock
+    InputStream mockInputStream;
+
+    @Mock
+    DateFormat mockDateFormat;
+
+    @Mock
+    AmazonAuthHelper mockAmazonAuth;
+
+    @Mock
+    Cursor mockCursor;
+
+    private MockWebServer mockWebServer;
+    private FormDataRepository formDataRepository;
+    private RestApi restApi;
+
+    @Before
+    public void setUp() throws IOException {
+        mockWebServer = new MockWebServer();
+        mockWebServer.start(8080);
```

Comment:
Do you actually start up a server or is it just setting the port and all the responses are mocked?

Reply:
it starts a server https://github.com/square/okhttp/blob/master/mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java#L344